domain/permission/state/state.go

```diff
@@ -195,25 +196,153 @@ func (s *State) UpsertPermission(ctx context.Context, args permission.UpsertPerm
 func (s *State) ReadUserAccessLevelForTarget(ctx context.Context, subject string, target corepermission.ID) (corepermission.Access, error) {
-	return corepermission.NoAccess, errors.NotImplementedf("ReadUserAccessLevelForTarget")
+	userAccessType := corepermission.NoAccess
+	db, err := s.DB()
+	if err != nil {
+		return userAccessType, errors.Trace(err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		userUUID, err := userUUIDForName(ctx, tx, subject)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		// Based on the grant to and grant from, find the permission,
+		// then the access type of it.
+		accessType, _, err := findAccessType(ctx, tx, target.Key, userUUID)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		userAccessType = corepermission.Access(accessType)
+		return nil
+	})
+	if err == nil {
+		return userAccessType, nil
+	}
+	return userAccessType, errors.Trace(domain.CoerceError(err))
 }
 
 // ReadAllUserAccessForUser returns a slice of the user access the given
 // subject's (user) has for any access type.
 func (s *State) ReadAllUserAccessForUser(ctx context.Context, subject string) ([]corepermission.UserAccess, error) {
-	return nil, errors.NotImplementedf("ReadAllUserAccessForUser")
+	db, err := s.DB()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	var (
+		permissions []readUserPermission
+		user        user
+	)
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		user, err = findUserByName(ctx, tx, subject)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		permissions, err = readUsersPermissions(ctx, tx, user.UUID)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		permissions, err = grantOnType(ctx, tx, permissions)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, errors.Trace(domain.CoerceError(err))
+	}
+
+	userAccess := make([]corepermission.UserAccess, len(permissions))
+	for i, p := range permissions {
+		userAccess[i] = p.toUserAccess(user)
+	}
+
+	return userAccess, nil
 }
 
 // ReadAllUserAccessForTarget return a slice of user access for all users
 // with access to the given target.
 func (s *State) ReadAllUserAccessForTarget(ctx context.Context, target corepermission.ID) ([]corepermission.UserAccess, error) {
-	return nil, errors.NotImplementedf("ReadAllUserAccessForTarget")
+	db, err := s.DB()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	var (
+		permissions []readUserPermission
+		users       map[string]user
+	)
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		// Get all permissions for target.Key
+		// Get all users from the list of permissions
+		// Combine data to return a slice of UserAccess.
+
+		permissions, err = targetPermissions(ctx, tx, target.Key)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		userUUIDs := make([]string, len(permissions))
+		for i, p := range permissions {
+			userUUIDs[i] = p.GrantTo
+		}
```

Comment:
Shouldn't this be a map? That way you can de-duplicate reoccurring grants.

Reply:
No, the permission table has a unique constraint that each combination of grant_on/grant_to. This method is returning all users (grant_to) who have a permission on a single target (grant_on). 

As well, if we used a map, and there were multiple combinations of grant_on/grant_to we'd get a random result.

Reply:
IN doesn't guarantee order in sqlite, you would need to add an order by to ensure that then.